getline.ts/src/blockchain.ts

```diff
@@ -263,21 +279,34 @@ export class GetlineBlockchain {
         const contract = this.web3.eth.contract(abi);
         const bytecode = await this.metabackend.getBytecode(contractName);
 
+        const gas: number = await (new Promise<number>((resolve, reject) => {
+            const fallback = 2000000;
+            const data = contract.getData(...params, { data: bytecode });
+            this.web3.eth.estimateGas({data}, (err: Error, cost: number) => {
+                if (err !== null) {
+                    logger(`Failed to estimate deployment gas consumption, falling back to default gas `
+                           + `(${fallback}): ${err}`);
+                    resolve(fallback);
+                    return;
+                }
+                logger(`Estimated gas consumption for contract deployment: ${cost}`);
+                resolve(cost);
+            });
+        }));
+
         return new Promise<Contract>((resolve, reject) => {
-            // TODO(q3k): Typify Web3.Contract.new
-            const contractAny: any = contract;
 
             const opts = {
                 data: bytecode,
                 from: this.web3.eth.coinbase,
-                gas: 4500000,
+                gas: Math.floor(gas * 1.1),
```

Comment:
Why

Reply:
Variations. Try running this code without the `* 1.1` and `* 2` and you'll find that more often than not your transactions will run out of gas. I haven't investigated the cause of it.

A better solution might be to add a static increase (instead of a relative one, like here). Or retry the transaction a couple of times with increasingly large gas limits, but that might be too slow to be worth the effort.

Reply:
Web3 gasEstimations are pretty good and add some leeway on top if I remember, are you sure you're estimating correctly? Especially for deployment which doesn't have any branching.

Reply:
It might be a Ganache issue..?

If I do `gas: gas + 100000` for mutations and `gas: gas + 1000` for deployments it seems to work fine in integration tests. This is still done using manual bisection, I'll gather some proper data in a second.

Reply:
When running with `gas: gas`:
```
gatherCollateral: gas estimate 113787
gatherCollateral: opts: {"gas":113787}
gatherCollateral: failed: Error: VM Exception while processing transaction: revert
```

When running the same test with `gas: gas + 100000`:

```
gatherCollateral: gas estimate 113787
gatherCollateral: opts: {"gas":213787}
gatherCollateral: tx hash 0x86f1cf643fc4b42d4f62a7fc75f67d7a02ee886476290eadbdc4bf9252b718f4
Receipt for 0x86f1cf643fc4b42d4f62a7fc75f67d7a02ee886476290eadbdc4bf9252b718f4: gasUsed: 113787
```
Hum :)